deadlines/src/main/java/com/palantir/deadlines/Deadlines.java

```diff
@@ -87,37 +87,77 @@ private static Optional<RemainingDeadline> getRemainingDeadlineInternal() {
+     * stored in deadline state), this method may throw a {@link DeadlineExpiredException} if, at the time of call,
+     * the selected deadline has already expired.
+     *
      * This function has no side-effects on the internal deadline state stored in a TraceLocal.
      *
      * @param proposedDeadline a proposed value for the deadline; the actual value used will be the minimum of
      * this value and one already set via a previous call to {@link #parseFromRequest}, if it exists
      * @param request the request object to write the encoding to
      * @param adapter a {@link RequestEncodingAdapter} that handles writing the header value to the request object
+     * @param enforced enforce deadline expiration on this call, and propagate enforcement via a header as well
      */
     public static <T> void encodeToRequest(
-            Duration proposedDeadline, T request, RequestEncodingAdapter<? super T> adapter) {
+            Duration proposedDeadline, T request, RequestEncodingAdapter<? super T> adapter, boolean enforced) {
         Optional<RemainingDeadline> deadlineFromState = getRemainingDeadlineInternal();
         long proposedDeadlineNanos = proposedDeadline.toNanos();
         if (deadlineFromState.isEmpty()) {
             // use proposedDeadline
-            checkExpiration(proposedDeadlineNanos, false);
+            checkExpiration(
+                    proposedDeadlineNanos,
+                    false,
+                    // we have no existing deadline in state, so the caller can dictate whether this
+                    // deadline should be enforced by receivers
+                    // this allows originators (the first to send a request with an Expect-Within header)
+                    // to control enforcement (and propagation of that enforcement) further along the call chain
+                    enforced);
             adapter.setHeader(
                     request, DeadlinesHttpHeaders.EXPECT_WITHIN, durationToHeaderValue(proposedDeadlineNanos));
+            if (enforced) {
+                adapter.setHeader(request, DeadlinesHttpHeaders.EXPECT_WITHIN_ENFORCED, "true");
+            }
         } else {
             // use the minimum of proposedDeadline and the one read from state
             RemainingDeadline stateDeadline = deadlineFromState.get();
+            // in cases where we do have an existing deadline in state, use the enforcement flag already stored,
+            // and ignore the parameter passed here.
+            // this means that even if callers _want_ to trigger enforcement, they cannot if they are propagating
+            // an existing deadline from an upstream RPC call chain (unless _that_ caller requested enforcement)
+            boolean enforcedFromState = stateDeadline.enforced();
+            if (enforcedFromState) {
+                adapter.setHeader(request, DeadlinesHttpHeaders.EXPECT_WITHIN_ENFORCED, "true");
+            }
```

Comment:
Wouldn't this only apply within the `else` branch below? If the proposed deadline is smaller than the state deadline, we'd use the input `enforced` parameter value

Reply:
maybe - i think it's perhaps debatable. In this case, we are selecting a lower deadline based on some local constraint (like a socket read timeout for this hop). But the fact that there was a deadline stored in state means that we previously received an `Expect-Within` header which may or may not have had the enforcement flag set. If it was received _without_ the flag, it may be surprising to a caller to have a request fail due to deadline expiration when they didn't explicitly ask for that.